storing-app-data/slides.md

```diff
@@ -0,0 +1,456 @@
+editor.apply();
+
+context.startActivity(MainScreenActivity.class);
+finish();
+```
+
+^ Here is example code for storing something via Shared Preferences. This is could that could very easily be part of your login process. But it has two bugs. If apply fails to write to disk, the user will be mysteriously logged out if you application is restarted. Additionally if the write is not immediate, the finish() will block the main thread on it completing.
+
+---
+
+## Even more important for legal and privacy status, such as a EULA or GDPR opt-out.
+
+^ Kicking your user out of the app due to an error is unpleasant, losing money is bad.
+
+---
+
+# [fit] Storage Performance
+
+---
+
+# Why is storage performance important?
+
+## The user experience should always be smooth.
+
+## Storage can cause frame drops.
+
+---
+
+# It is easy to drop frames
+
+Running at 60fps gives the device 16 milliseconds to render each frame.
+
+Even on a modern device, the p95 storage read latency is already in this neighborhood. On low end devices, this can be drastically slower.
+
+---
+
+# Storage Performance
+
+## Different types of data have different performance needs.
+
+---
+
+# Storage Performance
+
+If a cache read is latent, maybe it should be eagerly treated as a miss?
+
+If a cache write fails, does anyone care?
+
+^ Storage operations should be treated similarly to networking ops, they can similarly randomly fail or be latent.
+
+---
+
+# Storage Performance
+
+## Critical disk operations should block progress with a spinner until complete
+
+---
+
+# [fit] API Ergonomics
+
+---
+[.build-lists: true]
+
+# API Ergonomics
+
+* Must be understandable by juniors but sufficient for advanced users
+
+^ A junior needs to be able to use the API intuitively, but an advanced user needs to be able to tune the performance.
+
+* Uses a common async abstraction
+
+^ Rx is the de-facto standard for async behavior on Android, followed by ListenableFuture (especially in Google code). Callbacks are another common pattern in java. Whatever async API is chosen, it needs to be able to easily handle success and error state callbacks.
+
+* The right way to use it should be the easiest
+
+^ Doing the right thing must be the easy way to use it.
+
+* Consistant behavior for primitives and objects
+* Flexible on-disk format that is future proof
+
+^ If an API can be built so the calling thread does not matter, that is a great thing. Blocking the UI thread should require suppressing analysis.
+
+---
+
+# [fit] I want a key-value store that ensures all of this.
+
+---
+
+# [fit] I want a key-value store that ensures all of this.
+# So I built it. Twice.
+
+---
+
+# [fit] SimpleStore
+
+### Simple yet performant asynchronous file storage for Android
+
+---
+
+# SimpleStore
+
+https://github.com/uber/simple-store
+
+[.text: alignment(center)]
+
+---
+
+# Write a key
+
+```java
+try(SimpleStore simpleStore = SimpleStoreFactory.create(this, "<some-uuid-or-name>")) {
```

Comment:
Your demo code should be with Kotlin, it'll be more concise and easier to read, and most people folks will feel it adapts to their own model of usage better.